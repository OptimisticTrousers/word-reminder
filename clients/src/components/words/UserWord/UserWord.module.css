.user-word__word,
.user-word__panel {
  margin-top: 1rem;
}

.user-word__panel {
  border: 1px solid var(--border-color);
}

.word {
  background-color: var(--secondary-card-bkg-color);
  padding: 1rem;
  border-radius: 8px;
  border: 1px solid var(--border-color);
  transition: 0.4s;
}

.user-word__header {
  display: flex;
  justify-content: space-between;
  margin-bottom: 0.5rem;
}

.word__buttons {
  display: flex;
  align-items: center;
  gap: 0.75rem;
  margin-top: 0.75rem;
}

.word__buttons > * {
  flex: 1;
}

.word__button {
  display: flex;
  align-items: center;
  justify-content: center;
}

.word__button .word__icon {
  margin-right: 8px;
  width: 28px;
  height: 28px;
}

.word__form {
  display: flex;
  justify-content: center;
}

.word__form > * {
  flex: 1;
}

.user-word__panel {
  padding: 1rem;
  background-color: var(--accordion-bkg-color);
}
