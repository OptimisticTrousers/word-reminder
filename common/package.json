{
  "name": "common",
  "version": "1.1.2",
  "main": "./src/index.ts",
  "types": "./dist/index.d.ts",
  "scripts": {
    "coverage": "jest --coverage",
    "compile": "tsc",
    "build": "webpack --config webpack.prod.js",
    "postdev": "npm run compile",
    "dev": "webpack --config webpack.dev.js",
    "test": "jest"
  },
  "keywords": [],
  "author": "Tony Isern",
  "license": "ISC",
  "description": "A directory with common code and types used between the server, Chrome extension, and app",
  "devDependencies": {
    "@types/jest": "^29.5.14",
    "jest": "^29.7.0",
    "ts-jest": "^29.2.5",
    "ts-loader": "^9.5.2",
    "typescript": "^5.7.3",
    "webpack": "^5.97.1",
    "webpack-cli": "^6.0.1",
    "webpack-merge": "^6.0.1"
  }
}
