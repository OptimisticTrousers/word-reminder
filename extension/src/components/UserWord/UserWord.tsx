/* eslint-disable @typescript-eslint/no-explicit-any */
import CSSModules from "react-css-modules";
import styles from "./UserWord.module.css";
import { MdDelete } from "react-icons/md";
import { BsInfoCircleFill } from "react-icons/bs";
import { FC, useRef, useState } from "react";
import useHttp from "../../hooks/useHttp";
import { useMutation } from "@tanstack/react-query";
import { toast } from "react-toastify";
import { Word } from "..";
import { IUserWord } from "../../context/AuthContext";

const UserWord: FC<IUserWord> = CSSModules(
  (props) => {
    const accordion = useRef<HTMLDivElement | null>(null);
    const [isAccordionOpen, setIsAccordionOpen] = useState(false);
    const { remove } = useHttp();

    const handleAccordion = () => {
      setIsAccordionOpen((prevValue) => !prevValue);
      if (accordion.current) {
        accordion.current.style.maxHeight =
          accordion.current?.style.maxHeight !== "0px"
            ? `0px`
            : `${accordion.current?.scrollHeight}px`;
      }
    };

    const { data, status, error, mutate } = useMutation({
      mutationFn: () => {
        return remove(
          `${
            import.meta.env.VITE_API_DOMAIN
          }/users/665164760636f4834e053388/words/${props.word._id}`
        );
      },
      onSuccess: () => {
        toast.success("You have successfully deleted a word!");
      },
      onError: (error) => {
        console.error(error);
        toast.error("An unknown error occured while deleting a post.");
      },
    });

    const handleDelete = () => {
      mutate();
    };

    return (
      <div styleName="user-word">
        <article styleName="user-word__word word">
          <p styleName="word__name">{props.word.word}</p>
          <div styleName="word__divider"></div>
          <div styleName="word__buttons">
            <button
              styleName="word__button word__button--info"
              onClick={handleAccordion}
            >
              <BsInfoCircleFill styleName="word__icon" />
            </button>
            <button
              styleName="word__delete word__button--delete"
              onClick={handleDelete}
            >
              <MdDelete styleName="word__icon" />
            </button>
          </div>
        </article>
        <div
          ref={accordion}
          styleName={`user-word__panel ${
            isAccordionOpen && "user-word__panel--active"
          }`}
        >
          <Word {...props} />
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "ignore" }
);

export default UserWord;
