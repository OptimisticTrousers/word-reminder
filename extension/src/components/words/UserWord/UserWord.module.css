.user-word__word {
  padding: 18px;
  margin-top: 1rem;
}

.word {
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--color-word-bg);
  transition: 0.4s;
}

.word__divider {
  text-transform: uppercase;
  font-weight: 600;
  color: rgb(181, 181, 181);
  display: flex;
  flex-direction: column;
  margin: 0 25px;
  align-items: center;
  justify-content: center;
}

.word__divider::before,
.word__divider::after {
  background-color: #dbdbdb;
  content: "";
  display: block;
}

.word__divider::after {
  margin-left: 0;
  margin-top: 10px;
  flex: 1;
  width: 3px;
}

.word__divider::before {
  margin-bottom: 10px;
  margin-right: 0;
  flex: 1;
  width: 1px;
}

.word__buttons {
  display: flex;
}

.word__button--info {
  margin-right: 1rem;
}

.user-word__panel {
  padding: 0 18px;
  background-color: var(--color-accordion-bg);
  overflow: hidden;
  max-height: 0;
  transition: max-height 0.2s ease-out;
}

.user-word__panel--active {
  max-height: none;
  background-color: var(--color-active-bg);
}
