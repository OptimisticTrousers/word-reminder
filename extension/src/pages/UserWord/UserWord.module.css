.word__heading,
.word__learned,
.word__created-at,
.word__phonetic,
.word__phonetics,
.word__button,
.word__text,
.word__license,
.word__top,
.word__meanings,
.word__line,
.word__subtitle,
.word__definition,
.word__line,
.word__block,
.word__part-of-speech {
  margin-bottom: 0.5rem;
}

.word__buttons {
  display: flex;
  justify-content: space-between;
}

.word__button {
  display: flex;
  align-items: center;
}

.word__text {
  margin-right: 0.5rem;
}

.word__definition,
.word__block {
  padding: 1rem;
  border-radius: 5px;
  border: 1px solid var(--border-color);
  background-color: var(--primary-card-bkg-color);
}

.word__synonym,
.word__antonym,
.word__item,
.word__definition {
  list-style: none;
}

.word__span {
  font-weight: 600;
}
