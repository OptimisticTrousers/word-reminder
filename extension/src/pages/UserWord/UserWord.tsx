import {
  Definition,
  Detail,
  Meaning,
  Phonetic,
  User,
  UserWord as IUserWord,
  Word,
  Image,
} from "common";
import { Play, Trash } from "lucide-react";
import CSSModules from "react-css-modules";
import { Link, useOutletContext, useParams } from "react-router-dom";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { Loading } from "../../components/ui/Loading";
import { Error500 } from "../Error500";
import { userWordService } from "../../services/user_word_service";
import styles from "./UserWord.module.css";
import { ImageCarousel } from "../../components/ui/ImageCarousel";
import { FormEvent, useState } from "react";
import { DeleteUserWordModal } from "../../components/modals/DeleteUserWordModal";
import { useNotificationError } from "../../hooks/useNotificationError";

export const UserWord = CSSModules(
  function () {
    const { user }: { user: User } = useOutletContext();
    const { userWordId } = useParams();
    const userId = String(user.id);
    const queryClient = useQueryClient();

    const { data, isLoading, failureReason } = useQuery({
      queryKey: ["userWords", userWordId],
      queryFn: () => {
        return userWordService.getUserWord({
          userId,
          userWordId: String(userWordId),
        });
      },
    });
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
    const { showNotificationError } = useNotificationError();

    const { isPending, mutate } = useMutation({
      mutationFn: userWordService.updateUserWord,
      onSuccess: () => {
        queryClient.invalidateQueries({
          queryKey: ["userWords"],
        });
      },
      onError: showNotificationError,
    });

    function handleLearned(event: FormEvent<HTMLFormElement>) {
      event.preventDefault();
      const formData = new FormData(event.currentTarget);
      mutate({
        userId,
        userWordId: String(userWordId),
        body: {
          learned: (formData.get("learned") as string) === "true",
        },
      });
    }

    const disabled = isPending;

    if (failureReason) {
      return <Error500 message={failureReason.message} />;
    }

    if (isLoading) {
      return <Loading />;
    }

    function handleAudio(url: string) {
      const audio = new Audio(url);
      audio.play();
    }

    function toggleDeleteModal() {
      setIsDeleteModalOpen((prevIsDeleteModalOpen) => !prevIsDeleteModalOpen);
    }

    const json = data?.json;
    const userWord: (IUserWord & { word: Word; images: Image[] }) | undefined =
      json?.userWord;
    const details = userWord?.word.details;
    const images = userWord?.images ?? [];

    return (
      details && (
        <div styleName="word">
          {isDeleteModalOpen && (
            <DeleteUserWordModal
              toggleModal={toggleDeleteModal}
              userWordId={String(userWord.id)}
            />
          )}
          {<h2 styleName="word__heading">{details[0].word}</h2>}
          {
            <p styleName="word__learned">
              {userWord.learned ? "Learned word" : "Word not learned"}
            </p>
          }
          {
            <p styleName="word__created_at">
              {userWord.created_at.toLocaleString()}
            </p>
          }
          {<ImageCarousel images={images} />}
          <div styleName="word__buttons">
            <form styleName="word__form" onSubmit={handleLearned}>
              <input
                styleName="word__input word__input--hidden"
                type="hidden"
                id="learned"
                name="learned"
                required
                defaultValue={String(!userWord.learned)}
              />
              <button
                styleName="word__button word__button--learned"
                type="submit"
                disabled={disabled}
              >
                {disabled ? "Toggling Learned..." : "Toggle Learned"}
              </button>
            </form>
            <button
              styleName="word__delete word__button--delete"
              onClick={toggleDeleteModal}
              aria-haspopup={true}
              aria-label="Open delete user word modal"
            >
              <Trash styleName="word__icon" />
            </button>
          </div>
          {details.map((detail: Detail, index: number) => {
            return (
              <div key={index} styleName="word_content">
                <div styleName="word__top">
                  <div styleName="word__title">
                    {detail.phonetic && (
                      <p styleName="word__phonetic">{detail.phonetic}</p>
                    )}
                    <ul styleName="word__phonetics">
                      {detail.phonetics &&
                        detail.phonetics.map(
                          (phonetic: Phonetic, index: number) => {
                            return (
                              <li key={index} styleName="word__phonetic">
                                {phonetic.text && (
                                  <p styleName="word__text">{phonetic.text}</p>
                                )}
                                {phonetic.audio && (
                                  <button
                                    styleName="word__button"
                                    onClick={() => handleAudio(phonetic.audio!)}
                                  >
                                    <span styleName="word__text">
                                      Pronounce word
                                    </span>
                                    <Play styleName="word__icon" />
                                  </button>
                                )}
                                {phonetic.sourceUrl && (
                                  <Link
                                    to={phonetic.sourceUrl}
                                    styleName="word__link"
                                  >
                                    Phonetic Source
                                  </Link>
                                )}
                                {phonetic.license && (
                                  <div styleName="word__license">
                                    <p styleName="word__name">
                                      {phonetic.license.name}
                                    </p>
                                    <Link
                                      to={phonetic.license.url}
                                      styleName="word__url"
                                    >
                                      Phonetic License
                                    </Link>
                                  </div>
                                )}
                              </li>
                            );
                          }
                        )}
                    </ul>
                  </div>
                </div>
                <div styleName="word__container">
                  <div styleName="word__block">
                    <h3 styleName="word__subtitle">Origin</h3>
                    <p styleName="word__description">{detail.origin}</p>
                  </div>
                  {detail.meanings.map((meaning: Meaning, index: number) => (
                    <div styleName="word__meanings" key={index}>
                      <span styleName="word__part-of-speech">
                        {meaning.partOfSpeech} <hr styleName="word__line" />
                      </span>
                      <div styleName="word__block">
                        <h3 styleName="word__subtitle">Meanings</h3>
                        <ul styleName="word__definitions">
                          {meaning.definitions.map(
                            (definition: Definition, index: number) => (
                              <li styleName="word__definition" key={index}>
                                <p styleName="word__definition">
                                  {definition.definition}
                                </p>
                                {definition.example && (
                                  <p styleName="word__example">
                                    {definition.example}
                                  </p>
                                )}
                                <ul styleName="word__synonyms">
                                  {definition.synonyms &&
                                    definition.synonyms.map(
                                      (synonym: string, index: number) => {
                                        return (
                                          <li
                                            styleName="word__synonym"
                                            key={index}
                                          >
                                            {synonym}
                                          </li>
                                        );
                                      }
                                    )}
                                </ul>
                                <ul styleName="word__antonyms">
                                  {definition.antonyms &&
                                    definition.antonyms.map(
                                      (antonym: string, index: number) => {
                                        return (
                                          <li
                                            styleName="word__antonym"
                                            key={index}
                                          >
                                            {antonym}
                                          </li>
                                        );
                                      }
                                    )}
                                </ul>
                              </li>
                            )
                          )}
                        </ul>
                      </div>
                      {meaning.synonyms && (
                        <div styleName="word__block">
                          <h3 styleName="word__subtitle">Synonyms</h3>
                          <ul styleName="word__definitions">
                            {meaning.synonyms.map(
                              (synonym: string, index: number) => {
                                return (
                                  <li styleName="word__synonym" key={index}>
                                    {synonym}
                                  </li>
                                );
                              }
                            )}
                          </ul>
                        </div>
                      )}
                      {meaning.antonyms && (
                        <div styleName="word__block">
                          <h3 styleName="word__subtitle">Antonyms</h3>
                          <ul styleName="word__definitions">
                            {meaning.antonyms.map(
                              (antonym: string, index: number) => {
                                return (
                                  <li styleName="word__antonym" key={index}>
                                    {antonym}
                                  </li>
                                );
                              }
                            )}
                          </ul>
                        </div>
                      )}
                      {detail.license && (
                        <div styleName="word__license">
                          <p styleName="word__name">{detail.license.name}</p>
                          <Link to={detail.license.url} styleName="word__url">
                            Word License
                          </Link>
                        </div>
                      )}
                      <ul styleName="word__list">
                        {detail.sourceUrls &&
                          detail.sourceUrls.map(
                            (sourceUrl: string, index: number) => {
                              return (
                                <li key={index} styleName="word__item">
                                  <Link to={sourceUrl} styleName="word__link">
                                    Word Source
                                  </Link>
                                </li>
                              );
                            }
                          )}
                      </ul>
                    </div>
                  ))}
                </div>
              </div>
            );
          })}
        </div>
      )
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
