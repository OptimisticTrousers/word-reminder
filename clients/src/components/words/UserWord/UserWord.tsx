import { Detail, User } from "common";
import { GraduationCap, Info, Trash } from "lucide-react";
import { FormEvent, useState } from "react";
import CSSModules from "react-css-modules";
import { Link, useOutletContext } from "react-router-dom";

import { DeleteUserWordModal } from "../../modals/DeleteUserWordModal";
import { CondensedWord } from "../CondensedWord";
import styles from "./UserWord.module.css";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { userWordService } from "../../../services/user_word_service";
import { useNotificationError } from "../../../hooks/useNotificationError";

export const UserWord = CSSModules(
  function ({
    learned,
    created_at,
    updated_at,
    details,
    id,
  }: {
    learned: boolean;
    created_at: Date;
    updated_at: Date;
    details: Detail[];
    id: number;
  }) {
    const [isAccordionOpen, setIsAccordionOpen] = useState(false);
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);

    const queryClient = useQueryClient();
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const { showNotificationError } = useNotificationError();

    const { isPending, mutate } = useMutation({
      mutationFn: userWordService.updateUserWord,
      onSuccess: () => {
        queryClient.invalidateQueries({
          queryKey: ["userWords"],
        });
      },
      onError: showNotificationError,
    });

    const toggleAccordion = () => {
      setIsAccordionOpen((prevValue) => !prevValue);
    };

    function toggleDeleteModal() {
      setIsDeleteModalOpen((prevIsDeleteModalOpen) => !prevIsDeleteModalOpen);
    }

    function handleLearned(event: FormEvent<HTMLFormElement>) {
      event.preventDefault();
      const formData = new FormData(event.currentTarget);
      mutate({
        userId,
        userWordId: String(id),
        body: {
          learned: (formData.get("learned") as string) === "true",
        },
      });
    }

    const disabled = isPending;

    return (
      <>
        <div styleName="user-word">
          <article styleName="user-word__word word">
            <div styleName="user-word__header">
              <h3 styleName="word__name">{details[0].word}</h3>
              <Link to={`/userWords/${id}`}>More Word Details</Link>
            </div>
            <p styleName="word__name">
              Created At: {created_at.toLocaleString()}
            </p>
            <p styleName="word__name">
              Updated At: {updated_at.toLocaleString()}
            </p>
            <p styleName="word__name">Learned: {learned ? "Yes" : "No"}</p>
            <div styleName="word__buttons">
              <button
                id="accordion-button"
                type="button"
                styleName="word__button word__button--info"
                onClick={toggleAccordion}
                aria-expanded={isAccordionOpen}
                aria-label="More word details"
                aria-controls="accordion"
              >
                <Info styleName="word__icon" />
              </button>
              <form styleName="word__form" onSubmit={handleLearned}>
                <input
                  styleName="word__input"
                  type="hidden"
                  id="learned"
                  name="learned"
                  required
                  defaultValue={String(!learned)}
                />
                <button
                  styleName="word__button"
                  type="submit"
                  disabled={disabled}
                >
                  <GraduationCap styleName="word__icon" />
                  {disabled ? "Toggling Learned..." : "Toggle Learned"}
                </button>
              </form>
              <button
                styleName="word__delete"
                onClick={toggleDeleteModal}
                aria-haspopup={true}
                aria-label="Open delete user word modal"
              >
                <Trash styleName="word__icon" />
              </button>
            </div>
          </article>
          {isAccordionOpen && (
            <div
              id="accordion"
              styleName={`user-word__panel ${
                isAccordionOpen && "user-word__panel--active"
              }`}
              role="region"
              aria-labelledby="accordion-button"
            >
              <CondensedWord details={details} />
            </div>
          )}
        </div>
        {isDeleteModalOpen && (
          <DeleteUserWordModal
            toggleModal={toggleDeleteModal}
            userWordId={String(id)}
          />
        )}
      </>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
