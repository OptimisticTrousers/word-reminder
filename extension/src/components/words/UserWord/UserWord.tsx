import { Detail } from "common";
import { Info, Trash } from "lucide-react";
import { useState } from "react";
import CSSModules from "react-css-modules";
import { Link } from "react-router-dom";

import { DeleteUserWordModal } from "../../modals/DeleteUserWordModal";
import { CondensedWord } from "../CondensedWord";
import styles from "./UserWord.module.css";

export const UserWord = CSSModules(
  function ({
    learned,
    created_at,
    updated_at,
    details,
    id,
  }: {
    learned: boolean;
    created_at: Date;
    updated_at: Date;
    details: Detail[];
    id: number;
  }) {
    const [isAccordionOpen, setIsAccordionOpen] = useState(false);
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);

    const toggleAccordion = () => {
      setIsAccordionOpen((prevValue) => !prevValue);
    };

    function toggleDeleteModal() {
      setIsDeleteModalOpen((prevIsDeleteModalOpen) => !prevIsDeleteModalOpen);
    }

    return (
      <>
        <div styleName="user-word">
          <article styleName="user-word__word word">
            <p styleName="word__name">Word: {details[0].word}</p>
            <p styleName="word__name">
              Created At: {created_at.toLocaleString()}
            </p>
            <p styleName="word__name">
              Updated At: {updated_at.toLocaleString()}
            </p>
            <p styleName="word__name">Learned: {learned}</p>
            <div styleName="word__divider"></div>
            <div styleName="word__buttons">
              <button
                id="accordion-button"
                type="button"
                styleName="word__button word__button--info"
                onClick={toggleAccordion}
                aria-expanded={isAccordionOpen}
                aria-label="More word details"
                aria-controls="accordion"
              >
                <Info styleName="word__icon" />
              </button>
              <button
                styleName="word__delete word__button--delete"
                onClick={toggleDeleteModal}
                aria-haspopup={true}
                aria-label="Open delete user word modal"
              >
                <Trash styleName="word__icon" />
              </button>
              <Link to={`/userWords/${id}`}>More Word Details</Link>
            </div>
          </article>
          {isAccordionOpen && (
            <div
              id="accordion"
              styleName={`user-word__panel ${
                isAccordionOpen && "user-word__panel--active"
              }`}
              role="region"
              aria-labelledby="accordion-button"
            >
              <CondensedWord details={details} />
            </div>
          )}
        </div>
        {isDeleteModalOpen && (
          <DeleteUserWordModal
            toggleModal={toggleDeleteModal}
            userWordId={String(id)}
          />
        )}
      </>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
