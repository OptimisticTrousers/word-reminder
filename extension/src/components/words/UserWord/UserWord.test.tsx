import { Detail } from "common";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { UserWord } from "./UserWord";
import { MemoryRouter } from "react-router-dom";

vi.mock("../CondensedWord/CondensedWord", () => {
  return {
    CondensedWord: function ({ details }: { details: Detail[] }) {
      return (
        <>
          <div data-testid="details">{JSON.stringify(details)}</div>
        </>
      );
    },
  };
});

vi.mock("../../modals/DeleteUserWordModal", () => {
  return {
    DeleteUserWordModal: function () {
      return <div data-testid="modal"></div>;
    },
  };
});

const props = {
  id: "1",
  word_id: "1",
  user_id: "1",
  learned: false,
  updated_at: new Date("December 17, 1995 03:24:00"),
  created_at: new Date("December 17, 1995 03:24:00"),
  details: [{ word: "exemplary", phonetics: [], meanings: [] }],
};

describe("UserWord component", () => {
  it("does not renders the word by default", async () => {
    render(
      <MemoryRouter>
        <UserWord {...props} />
      </MemoryRouter>
    );

    const details = screen.queryByTestId("details");
    expect(details).not.toBeInTheDocument();
  });

  it("it opens delete modal", async () => {
    render(
      <MemoryRouter>
        <UserWord {...props} />
      </MemoryRouter>
    );
    const user = userEvent.setup();

    const deleteButton = screen.getByRole("button", {
      name: "Open delete user word modal",
    });
    await user.click(deleteButton);

    const deleteModal = screen.getByTestId("modal");
    expect(deleteModal).toBeInTheDocument();
  });

  it("it opens the accordion", async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <UserWord {...props} />
      </MemoryRouter>
    );
    const user = userEvent.setup();

    const infoButton = screen.getByRole("button", {
      name: "More word details",
    });
    await user.click(infoButton);

    const details = screen.getByTestId("details");
    expect(details).toHaveTextContent(JSON.stringify(props.details));
    expect(asFragment()).toMatchSnapshot();
  });

  it("it closes the accordion", async () => {
    render(
      <MemoryRouter>
        <UserWord {...props} />
      </MemoryRouter>
    );
    const user = userEvent.setup();

    const infoButton = screen.getByRole("button", {
      name: "More word details",
    });
    await user.click(infoButton);
    await user.click(infoButton);

    const details = screen.queryByTestId("details");
    expect(details).not.toBeInTheDocument();
  });
});
