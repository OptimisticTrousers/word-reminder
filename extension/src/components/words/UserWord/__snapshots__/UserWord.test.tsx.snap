// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserWord component > it opens the accordion 1`] = `
<DocumentFragment>
  <div
    class="_user-word_a41cd7"
  >
    <article
      class="_user-word__word_a41cd7 _word_a41cd7"
    >
      <p
        class="_word__name_a41cd7"
      >
        Word: exemplary
      </p>
      <p
        class="_word__name_a41cd7"
      >
        Created At: 12/17/1995, 3:24:00 AM
      </p>
      <p
        class="_word__name_a41cd7"
      >
        Updated At: 12/17/1995, 3:24:00 AM
      </p>
      <p
        class="_word__name_a41cd7"
      >
        Learned: 
      </p>
      <div
        class="_word__divider_a41cd7"
      />
      <div
        class="_word__buttons_a41cd7"
      >
        <h3
          class="_word__heading_a41cd7"
        >
          <button
            aria-controls="accordion"
            aria-expanded="true"
            aria-label="More word details"
            class="_word__button_a41cd7 _word__button--info_a41cd7"
            id="accordion-button"
            type="button"
          >
            <svg
              class="lucide lucide-info _word__icon_a41cd7"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <path
                d="M12 16v-4"
              />
              <path
                d="M12 8h.01"
              />
            </svg>
          </button>
        </h3>
        <h3
          class="_word__heading_a41cd7"
        >
          <button
            aria-haspopup="true"
            aria-label="Open delete user word modal"
            class="_word__delete_a41cd7 _word__button--delete_a41cd7"
          >
            <svg
              class="lucide lucide-trash _word__icon_a41cd7"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 6h18"
              />
              <path
                d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
              />
              <path
                d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
              />
            </svg>
          </button>
        </h3>
        <a
          data-discover="true"
          href="/userWords/1"
        >
          More Word Details
        </a>
      </div>
    </article>
    <div
      aria-labelledby="accordion-button"
      class="_user-word__panel_a41cd7 _user-word__panel--active_a41cd7"
      id="accordion"
      role="region"
    >
      <div
        data-testid="details"
      >
        [{"word":"exemplary","phonetics":[],"meanings":[]}]
      </div>
    </div>
  </div>
</DocumentFragment>
`;
